src/main/kotlin/de/cyface/collector/handler/MeasurementHandler.kt

```diff
@@ -55,194 +55,185 @@ import java.util.UUID
+            session.destroy()
+            ctx.fail(ENTITY_UNPARSABLE, e)
         } catch (e: RuntimeException) {
             ctx.fail(e)
         }
     }
 
     /**
-     * Handles the chunk upload request when there was no chunk file found on the server yet.
+     * Called if checking the data and loading it either to temporary storage or to the final location, has finished
+     * successfully.
      *
-     * @param ctx The Vertx `RoutingContext` used to write the response
-     * @param request the request to read the body from
-     * @param session the session which was passed with the request
-     * @param user the user which was authenticated to this request
-     * @param contentRange the content range information from the header
-     * @param metaData the metadata from the request header
-     * @param path `String` if the session contained a path to the chunk file or `null` otherwise.
+     * @param status: The return status of the check.
+     * @param context: The `RoutingContext` used by the current request.
+     * @param session: The current HTTP session.
      */
-    private fun handleFirstChunkUpload(
-        ctx: RoutingContext,
-        request: HttpServerRequest,
-        session: Session,
-        user: User,
-        contentRange: ContentRange,
-        metaData: RequestMetaData,
-        path: String?
-    ) {
-        if (contentRange.fromIndex != "0") {
-            if (path == null) {
-                // I.e. the server received data in a previous request but now cannot find the `path` session value.
-                // Unsure when this can happen. Not accepting the data. Asking to restart upload (`404`).
-                LOGGER.warn(String.format("Response: 404, path is null and unexpected content range: %s", contentRange))
-            } else {
-                // Server received data in a previous request but the chunk file was probably cleaned by cleaner task.
-                // Can't return 308. Google API client lib throws `Preconditions` on subsequent chunk upload.
-                // This makes sense as the server reported before that bytes (>0) were received.
-                LOGGER.warn(String.format("Response: 404, Unexpected content range: %s", contentRange))
+    private fun onCheckSuccessful(status: Status, context: RoutingContext, session: Session) {
+        when (status.type) {
+            StatusType.INCOMPLETE -> {
+                val byteSize = status.byteSize
+                val range = String.format(Locale.ENGLISH, "bytes=0-%d", byteSize - 1)
+                context.response().putHeader("Range", range)
+                context.response().putHeader("Content-Length", "0")
+                context.response().setStatusCode(RESUME_INCOMPLETE).end()
+            }
+            StatusType.COMPLETE -> {
+                // not removing session
+                session.remove<Any>(UPLOAD_PATH_FIELD)
+                storageService.clean(status.uploadIdentifier).onSuccess {
```

Comment:
The comment "NOT removing session" is contrary to the next line, which removes the session. But this was the case before, too:

```
uploadFuture.onSuccess { result: String? ->
    // Not removing session to allow the client to check the upload status if interrupted
    LOGGER.debug("Response: 201")
    clean(ctx.session(), measurement.binary)
    ctx.response().setStatusCode(201).end()
```

I guess the cleanest way to fix this and to ensure the server supports this use-case (which is definitely possible) is to write a test which uploads a file successfully, but then sends a status request afterwards (assuming, that the 201 CREATED response from the server did not arrive at the client as that response was interrupted, but the upload itself was successful).

Then, when you fix this (by probably not removing the session) ensure the comment also contains the reason why we do not remove the session after a successful upload:
`// Not removing session to allow the client to check the upload status if interrupted`

Reply:
This is not clear to me. I guess we should remove the session here, shouldn't we? Asking for the status of a finished session should simply return, that that session does not exist.